packages/block-editor/src/components/block-controls/index.js

```diff
@@ -48,6 +54,9 @@ function BlockControlsFill( { controls, children } ) {
 	);
 }
 
+const buildSlotName = ( isExpanded ) =>
+	`BlockControls${ isExpanded ? '-expanded' : '' }`;
```

Comment:
I don't understand why you had to use a different slot name though?

Reply:
I was suggesting to use the exact same slot and fills but have some logic on the wrapper components to avoid rendering if there is already an expanded fill on the slot or something like that. Would that be possible?

Reply:
How else would I make sure that expanded fills don't land inside the default content OR that non-expanded ones replace the entire content?

Reply:
I tried inspecting the list of fills and use the `isExpanded` prop, but then I realized it's just reimplementing the slot name.

Reply:
Something's wrong with GitHub, I only saw your second comment a few minutes after posting mine.

> I was suggesting to use the exact same slot and fills but have some logic on the wrapper components to avoid rendering if there is already an expanded fill on the slot or something like that. Would that be possible?

I agree this would be the perfect solution. At the moment accessing fills outside of the slot is not possible and some refactoring would be required first so I went for the "wrapping slot" approach. Let me see what would it take to make it happen, it seems like virtually bubbling slots have a hook so maybe it would be easy.

Reply:
Yes, I'm not sure either whether it's better or not, the current implementation is also decent. It's just that it feels more conceptually correct. (shouldn't block this PR though)

Reply:
A small refactoring of bubbles-virtually/fill would make it possible to detect the presence of expanded fills outside of slot:

**Before:**
https://github.com/WordPress/gutenberg/blob/9f2ca8bd6bf1f685b921ff9f31dcfc9311446fd6/packages/components/src/slot-fill/bubbles-virtually/fill.js#L16-L18

**After:**
```js
export default function Fill( { name, children, refProps } ) {
	const slot = useSlot( name );
	const ref = useRef( { rerender: useForceUpdate(), ...refProps } );
```

With that in place, the following is possible:

```js
	const { fills } = useSlot( 'BlockControls' );
	const expandedFills = fills.filter( ( ref ) => ref.current?.isExpanded  );
```

So far so good - so we can use different logic for both cases. The problem is rendering only expanded or non-expanded fills here:

```js
	<BlockControls.Slot
		bubblesVirtually
		className="block-editor-block-toolbar__slot"
	/>
```

We don't have a granular control over fills at the slot level - `bubblesVirtually` fills render themselves using portals. So either:

* All BlockControls fills would have to keep track of all other BlockControl fills and handle that internally (sounds awful), OR
* Expanded and non-expanded fills would have to create portals with different target element, for example based on slot name

So we're back to square one in regard to using slots and fills names. There would be a small win here in that we could get rid of the wrapping slot and make it a sibling of the default content instead, but it would still be a separate slot.